Update warp.py

gkimbell opened the pull request:
## Description
Changes proposed to address bug reported in https://github.com/qgis/QGIS/issues/31276. The change to line 218 addresses the case where an output extent is defined but its CRS is left blank (implicit). The change to line 220 ensures that a USER: coordinate system is translated into a Proj4 string that is recognised by GDAL.

Apologies if this pull request doesn't comply with some items on the checklist (I'm unfamiliar with this process)

Fixes #31276

## Checklist

<!-- Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.
-->

- [ ] Commit messages are descriptive and explain the rationale for changes
- [ ] Commits which fix bugs include `Fixes #11111` at the bottom of the commit message
- [ ] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [ ] New unit tests have been added for core changes
- [ ] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit
- [ ] I have evaluated whether it is appropriate for this PR to be backported, backport requests are left as label or comment

gkimbell committed:
Update warp.py

Changes suggested to address bug reported in https://github.com/qgis/QGIS/issues/31276

roya0045 commented:
Have you tested this?

I assume that the best way to ensure that this work would be to make a test in there https://github.com/qgis/QGIS/blob/master/python/plugins/processing/tests/GdalAlgorithmsRasterTest.py

nyalldawson commented:
Fix looks good!

As @roya0045 has pointed out, we also need a unit test covering this. It would slot in right near these similar tests: https://github.com/qgis/QGIS/blob/master/python/plugins/processing/tests/GdalAlgorithmsRasterTest.py#L1638

gkimbell committed:
Update GdalAlgorithmsRasterTest.py

Unit test added at lines 1662-1674 to address case where gdalwarp is run with a user-defined target extent and a custom projection. Relates to bug report #31276 and pull request #31320

roya0045 commented on python/plugins/processing/tests/GdalAlgorithmsRasterTest.py:
```diff
@@ -1659,6 +1659,20 @@ def testWarp(self):
+                 '-s_srs "+proj=utm +zone=36 +south +a=63785 +b=6357 +towgs84=-143,-90,-294,0,0,0,0 +units=m +no_defs" -t_srs "+proj=utm +zone=36 +south +a=63785 +b=6357 +towgs84=-143,-90,-294,0,0,0,0 +units=m +no_defs" -r near -te 18.67 45.78 18.7 45.81 -te_srs "+proj=utm +zone=36 +south +a=63785 +b=6357 +towgs84=-143,-90,-294,0,0,0,0 +units=m +no_defs" -of JPEG ' +
+                 source + ' ' +
+                 outdir + '/check.jpg'])
+
```
I would suggest you to create a distinct test. Basically copy this function, and remove elements that belong to the other test and validate elements related to your changes. This way we can more easily monitor when something breaks this functionality.

gkimbell committed:
Update GdalAlgorithmsRasterTest.py

Custom crs changed (to geographic) to distinguish the test from the previous one

gkimbell committed:
Update GdalAlgorithmsRasterTest.py

Renamed custom crs variable

gkimbell commented:
I've added a tentative unit test at lines 1662-1674 of https://github.com/gkimbell/QGIS/blob/master/python/plugins/processing/tests/GdalAlgorithmsRasterTest.py#L1662 

Not too confident about this as I have been unable to run the unit test properly on my PC (combination of trying to do things in Windows and extreme incompetence). Running an equivalent test manually in QGIS Desktop works OK after the edits to warp.py.

roya0045 commented:
Since there are so many tests in this classes it should be fine, I didn't think that all the warp parameters were tested in the same function. This should be fine.

roya0045 commented:
@gkimbell could you do a minimal commit to trigger to force travis to build and test again. I don't think your changes are what is causing the test to fail. If you force travis to rebuild well be able to know if everything is ok and the devs will be able to merge it once they get the green check mark.

gkimbell committed:
Update GdalAlgorithmsRasterTest.py

New commit to trigger Travis (output format of new unit test changed to geotiff)

gkimbell committed:
Update GdalAlgorithmsRasterTest.py

Correction to output file in console command
